pkg/pipeline/common_steps.go

```diff
@@ -0,0 +1,166 @@
+package pipeline
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/projectsyn/lieutenant-operator/pkg/apis"
+	synv1alpha1 "github.com/projectsyn/lieutenant-operator/pkg/apis/syn/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func getDefaultDeletionPolicy() synv1alpha1.DeletionPolicy {
+	policy := synv1alpha1.DeletionPolicy(os.Getenv("DEFAULT_DELETION_POLICY"))
+	switch policy {
+	case synv1alpha1.ArchivePolicy, synv1alpha1.DeletePolicy, synv1alpha1.RetainPolicy:
+		return policy
+	default:
+		return synv1alpha1.ArchivePolicy
+	}
+}
+
+func addDeletionProtection(instance PipelineObject, data *ExecutionContext) ExecutionResult {
+
+	if data.Deleted {
+		return ExecutionResult{}
+	}
+
+	config := os.Getenv(protectionSettingEnvVar)
+
+	protected, err := strconv.ParseBool(config)
+	if err != nil {
+		protected = true
+	}
+
+	if protected {
+		annotations := instance.GetObjectMeta().GetAnnotations()
+
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+
+		if _, ok := annotations[DeleteProtectionAnnotation]; !ok {
+			annotations[DeleteProtectionAnnotation] = "true"
+		}
+
+		instance.GetObjectMeta().SetAnnotations(annotations)
+	}
+
+	return ExecutionResult{}
+
+}
+
+// addTenantLabel adds the tenant label to an object.
+func addTenantLabel(obj PipelineObject, data *ExecutionContext) ExecutionResult {
+	labels := obj.GetObjectMeta().GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	if labels[apis.LabelNameTenant] != obj.GetTenantRef().Name {
+		labels[apis.LabelNameTenant] = obj.GetTenantRef().Name
+	}
+
+	obj.GetObjectMeta().SetLabels(labels)
+
+	return ExecutionResult{}
+}
+
+func updateObject(obj PipelineObject, data *ExecutionContext) ExecutionResult {
+
+	resourceVersion := obj.GetObjectMeta().GetResourceVersion()
+
+	rtObj, ok := obj.(runtime.Object)
+	if !ok {
+		return ExecutionResult{
+			Abort: true,
+			Err:   fmt.Errorf("object ist not a valid runtime.object: %v", obj.GetObjectMeta().GetName()),
+		}
+	}
+
+	err := data.Client.Update(context.TODO(), rtObj)
+	if err != nil {
+		return ExecutionResult{Err: err}
+	}
+
+	// Updating the status if either there were changes or the object is deleted will
+	// lead to some race conditions. By checking first we can avoid them.
+	if resourceVersion == obj.GetObjectMeta().GetResourceVersion() && !data.Deleted {
```

Comment:
Might be more clear:
```suggestion
	if resourceVersion == rtObj.GetObjectMeta().GetResourceVersion() && !data.Deleted {
```
The call to `Update()` will always increment the `ResourceVersion` so the status is never updated?

Reply:
According to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency

it's only increased when the object is modified.

Reply:
> Might be more clear:

The `rtObj runtime.Object` does not expose the ObjectMetada. So it can't be used in there.

Reply:
Ah, I see. Well then 👌